### PR TITLE
Build docs updates

### PR DESCRIPTION
**Pull request by robbrockbank**
Update build docs to include details about more make targets, and details on how to run fast and single STs, start up consul and etcd, installing pycalico library etc.

### CODE REVIEW

**Comment by tomdee** [comment]
Might be good for you to rebase in light of the libnetwork split

File: docs/Building.md [hunk 1]
```diff
@@ -1,28 +1,173 @@
-# Building Binaries
+# Building and testing Calico Docker
 
-You will require 2 binary builds to use the Calico Docker Prototype:
+This document describes how to build the Calico Docker binaries and how to run the test suites.
```
Comment (tomdee) [inline docs/Building.md]: I don't like the term calico docker binaries

File: docs/Building.md [hunk 2]
```diff
@@ -1,28 +1,173 @@
...
+
+
+## Building Binaries and running the test suites
```
Comment (tomdee) [inline docs/Building.md]: Again with the binaries

File: docs/Building.md [hunk 3]
```diff
@@ -1,28 +1,173 @@
...
+
+You will require 2 binary builds to use Calico Docker:
```
Comment (tomdee) [inline docs/Building.md]: I think you can call calicoctl a binary but calling a docker image a binary is a bit odd.

File: docs/Building.md [hunk 4]
```diff
@@ -1,28 +1,173 @@
...
-## Building the Docker Image
+The following sections describe how to build the binaries, and how to run the Unit Test
+and System Test suites.
 
- From the root directory of the checked out repository, run
+### Building the 'calico/node' Docker Image
+
+From the root directory of the checked out repository, run
 
     sudo make node
```
Comment (tomdee) [inline docs/Building.md]: This doesn't need sudo does it?

File: docs/Building.md [hunk 5]
```diff
@@ -1,28 +1,173 @@
...
 
 This builds the Dockerfile giving it the name `calico/node:latest`, which is the default image name `calicoctl` will look for when starting Calico services.
```
Comment (tomdee) [inline docs/Building.md]: ...the default image name _that the master branch version of_ calicoctl will look for

File: docs/Building.md [hunk 6]
```diff
@@ -1,28 +1,173 @@
...
 
-## Building `calicoctl` CLI Tool - mainline
+### Building `calicoctl` CLI Tool
+
+Our mainline makefile target uses a Debian-based container to build the `calicoctl` tool.  If this is not appropriate
```
Comment (tomdee) [inline docs/Building.md]: mainline? What do you mean by that?

File: docs/Building.md [hunk 7]
```diff
@@ -1,28 +1,173 @@
...
+for your Linux distribution, you can build the tool in a python virtualenv on your distro.  Follow the appropriate
+instructions below.
+
+#### Mainline build (Debian-based container build)
+
+From the root directory of the checked out repository, run
 
     sudo make binary
```
Comment (tomdee) [inline docs/Building.md]: Are you sure you need sudo?

File: docs/Building.md [hunk 8]
```diff
@@ -1,28 +1,173 @@
...
+
+    pyinstaller calico_containers/calicoctl.py -a -F -s --clean
+
+This builds `calicoctl` command line tool in the `dist/` directory.
+
+
+## The Calico Docker test suites and test environment
+
+There are two test suites used to validate Calico Docker function - unit tests and system tests.  In addition to the
+tests, there make targets to spin up an instance of etcd and consul - useful for test deployments.
```
Comment (tomdee) [inline docs/Building.md]: typo

Comment (tomdee) [inline docs/Building.md]: I would say manual testing, rather than test deployments

File: docs/Building.md [hunk 9]
```diff
@@ -1,28 +1,173 @@
...
+
+If you are developing Calico Docker code that you would like to contribute upstream, please ensure at a minimum that
+the unit tests and system tests successfully run, and, preferably, ensure new functionality is unit tested and system
+tested where necessary.  We aim to keep improving our total code and path coverage with each check-in.
+
+### Spinning up an instance of etcd and consul
```
Comment (tomdee) [inline docs/Building.md]: This feels a bit wordy and maybe over prominent. You've already called out these things above and say why you would use them. All that's left is to say what they're called (which you could just do above) and maybe mention the IP override thing.

File: docs/Building.md [hunk 10]
```diff
@@ -1,28 +1,173 @@
...
+
+This starts an consul server bound to a local IP address (automatically determined).  If you have multiple local IP
+addresses and wish to explicitly select an IP address to bind to, set the environment variable LOCAL_IP_ENV with the
+correct IP address.  For example:
+
+    LOCAL_IP_ENV=192.168.0.23 make run-etcd
+
+### Running the unit tests
+
+The Calico Docker unit tests provide code and path coverage of the main pycalico modules, a number of the integration
```
Comment (tomdee) [inline docs/Building.md]: Nope - some of these tests are moved to libcalico so I think this is wrong. And we'll be moving the other stuff soon so best to just keep this a bit less specific.

File: docs/Building.md [hunk 11]
```diff
@@ -1,28 +1,173 @@
...
+and plugin modules, and some of the calicoctl command line tool processing.
 
-We use a Debian-based container to build the `calicoctl` tool above.  If that is not appropriate to your Linux distribution, you can build in a python virtualenv on your distro.
+To run the unit tests, start by setting up a virtual env using the instructions detailed
+[here](#setting-up-a-virtualenv).
 
-Set up a [Python Virtual Environment](http://docs.python-guide.org/en/latest/dev/virtualenvs/)
+From the root directory of the checked out repository, run
+
+  make ut
```
Comment (tomdee) [inline docs/Building.md]: This isn't being rendered properly as code

File: docs/Building.md [hunk 12]
```diff
@@ -1,28 +1,173 @@
...
+
+    LOCAL_IP_ENV=192.168.0.23 make run-etcd
+
+### Running the unit tests
+
+The Calico Docker unit tests provide code and path coverage of the main pycalico modules, a number of the integration
+and plugin modules, and some of the calicoctl command line tool processing.
 
-We use a Debian-based container to build the `calicoctl` tool above.  If that is not appropriate to your Linux distribution, you can build in a python virtualenv on your distro.
+To run the unit tests, start by setting up a virtual env using the instructions detailed
```
Comment (tomdee) [inline docs/Building.md]: You don't need a venv to run the UTs...

File: docs/Building.md [hunk 13]
```diff
@@ -1,28 +1,173 @@
...
+multiple local IP addresses and wish to explicitly select an IP address to bind to, set the environment variable
+LOCAL_IP_ENV with the correct IP address.  For example:
+
+    sudo LOCAL_IP_ENV=192.168.0.23 make st
+
+The STs require root access to run, so they run with sudo.
+
+#### The full ST suite
+
+To run the unit tests, start by setting up a virtual env using the instructions detailed
```
Comment (tomdee) [inline docs/Building.md]: typo - unit

File: docs/Building.md [hunk 14]
```diff
@@ -1,28 +1,173 @@
...
+[here](#setting-up-a-virtualenv).
+
+From the root directory of the checked out repository, run
+
+    sudo make st
```
Comment (tomdee) [inline docs/Building.md]: I don't think the venv instructions are compatible with sudo.

File: docs/Building.md [hunk 15]
```diff
@@ -1,28 +1,173 @@
...
+To run the unit tests, start by setting up a virtual env using the instructions detailed
+[here](#setting-up-a-virtualenv).
+
+From the root directory of the checked out repository, run
+
+    sudo make fast-st
+
+Unlike the full ST suite, the fast ST suite does not build the calicoctl binary - instead it runs calicoctl directly
+from the python intepreter.  It starts containers running etcd and consul, and then runs all of the tests defined in
+tests/st that do not have the "slow" attrbiute assigned to them.
```
Comment (tomdee) [inline docs/Building.md]: typo

File: docs/Building.md (hunk 13 above)
Comment (tomdee) [inline docs/Building.md]: I think virtual env is one word

File: docs/Building.md [hunk 16]
```diff
@@ -1,28 +1,173 @@
...
+
+#### Running a subset of system tests
+
+If you wish to test a single system test module (particularly useful when writing a new system test, or running a
+quiock sanity check), use the full ST make target and set the environment variable ST_TO_RUN with the path of the module
```
Comment (tomdee) [inline docs/Building.md]: typo

Comment (tomdee) [inline docs/Building.md]: They don't need to run the full ST make targer

File: docs/Building.md [hunk 17]
```diff
@@ -1,28 +1,173 @@
...
+To run the single no-orchestrator, mainline multi-host test:
+
+    sudo ST_TO_RUN=calico_containers/tests/st/no_orchestrator/test_mainline_multi_host.py make st
+
+
+## Setting up a virtualenv
+
+A [Python Virtual Environment](http://docs.python-guide.org/en/latest/dev/virtualenvs/) can be used for building
+and running the Calico Docker test suites.  To create a virtualenv, from the root directory of the checked out
+repository, run
```
Comment (tomdee) [inline docs/Building.md]: Isn't there a make target for this already...

**Comment by tomdee** [comment]
OK, done for now. Can I take another look at this when you've made the markups

File: build_calicoctl/requirements.txt [hunk 18]
```diff
@@ -15,3 +15,5 @@ netaddr==0.7.15
 flask
 gunicorn
 subprocess32
+git+https://github.com/Metaswitch/python-etcd.git
```
Comment (alexwlchan) [inline build_calicoctl/requirements.txt]: The python-etcd repo has also been migrated to the projectcalico organisation.

File: docs/Building.md (hunk 3 above)
Comment (alexwlchan) [inline docs/Building.md]: [Nitpick] I would spell out the number; I think that’s more common in technical documentation/formal writing.

File: docs/Building.md [hunk 19]
```diff
@@ -1,28 +1,173 @@
...
+From the root directory of the checked out repository, run
 
     sudo make binary
 
 This builds `calicoctl` command line tool in the `dist/` directory.
 
-## Building `calicoctl` CLI Tool - virtualenv
+#### Virtual environment build
+
+Setup a virtualenv using the instructions detailed [here](#setting-up-a-virtualenv).
```
Comment (alexwlchan) [inline docs/Building.md]: Rather than linking with a generic “here”, it would be better to give the link some descriptive text like “using the instructions below”.

File: docs/Building.md (hunk 8 above)
Comment (alexwlchan) [inline docs/Building.md]: I believe “Consul” is supposed to be capitalised.

File: docs/Building.md [hunk 20]
```diff
@@ -1,28 +1,173 @@
...
+
+To run the unit tests, start by setting up a virtual env using the instructions detailed
+[here](#setting-up-a-virtualenv).
+
+From the root directory of the checked out repository, run
+
+    sudo make fast-st
+
+Unlike the full ST suite, the fast ST suite does not build the calicoctl binary - instead it runs calicoctl directly
+from the python intepreter.  It starts containers running etcd and consul, and then runs all of the tests defined in
```
Comment (alexwlchan) [inline docs/Building.md]: “Python” should be capitalised.

File: docs/Building.md [hunk 21]
```diff
@@ -1,28 +1,173 @@
...
+    sudo ST_TO_RUN=calico_containers/tests/st/bgp make st
+
+To run the single no-orchestrator, mainline multi-host test:
+
+    sudo ST_TO_RUN=calico_containers/tests/st/no_orchestrator/test_mainline_multi_host.py make st
+
+
+## Setting up a virtualenv
+
+A [Python Virtual Environment](http://docs.python-guide.org/en/latest/dev/virtualenvs/) can be used for building
```
Comment (alexwlchan) [inline docs/Building.md]: I would amend this to read

> A Python virtual environment, *or virtualenv*,

to explain the use of the abbreviation throughout this doc.

File: docs/Building.md (hunk 12 above)
Comment (robbrockbank) [inline docs/Building.md]: i thought we needed nose, mock, and a few other things to be installed?

Comment (robbrockbank) [inline docs/Building.md]: Ah  no --- we use a container.  fair'nuf

**Comment by tomdee** [comment]
LGTM
